Supplement owner check with access check for --install-devices

juergbi opened the pull request:
The README claims that any user will be able to run commands as well as `--install-devices` and `--uninstall-devices`. However, that's not actually the case. The latter two are restricted to the owner of the chroot.

It makes sense to guard the creation of directory entries in chroot `/dev` directories. However, having read and write access to the `/dev` directory should suffice. This allows the chroot owner to control which other users can install devices, if any.

I'd suggest merging this after #14 as that untangles the `/dev/shm` mount from device node creation on Linux.

juergbi committed:
Supplement owner check with access check for --install-devices

It makes sense to guard the creation of directory entries in chroot /dev
directories. However, having read and write access to the /dev directory
should suffice. This allows the chroot owner to control which other
users can install devices, if any.

jeremiahbonney commented on userchroot.c:
```diff
@@ -566,11 +566,28 @@ int main(int argc, char* argv[], char* envp[]) {
+
+    // this mode can only be run by users with write access to the
+    // /dev directory in the chroot image.
+    if (access(dev_path, R_OK | W_OK | X_OK) != 0) {
```
In addition to doing this, does it make sense to always allow the owner of the chroot to call install/uninstall devices? I can't quite imagine a case in practice where the owner of the chroot wouldn't have write permissions to `/dev`, but if such a case came up that's the behavior I'd expect.

juergbi replied:
If this was a new tool, I would probably not add such an exception unless there was a clear use case. However, to reduce the risk of breaking scenarios for current users of `userchroot` it sounds sensible.

If we make sure that `/dev` is not a symlink, always allowing the chroot owner to install/uninstall devices should be ok. If `/dev` was a symlink, the chroot owner could get `userchroot` to manipulate the wrong directory.

Side-note: All these path checks before the real operations are theoretically susceptible to small TOCTOU race conditions. However, that's already the case in master, so those should be fixed in a separate PR if considered significant. Directory fds and *at() syscalls should help if they are supported by all supported platforms.

juergbi replied:
I've updated the branch with the `/dev` symlink check and allowing chroot owners to install devices even if they don't have write access to `/dev`.

jeremiahbonney reviewed (commented):
Thanks for the PR, sorry I've been so slow to respond. One minor comment inline.

juergbi committed:
Check /dev is not a symlink for --install-devices

By creating /dev as a symlink the chroot owner could create or remove
the device nodes in arbitrary directories.

jeremiahbonney reviewed (approved):
The symlink check for `dev` is a great addition, thanks for the PR! LGTM
